core/uploadDiveLogsDE.cpp

```diff
@@ -140,7 +138,7 @@ bool uploadDiveLogsDE::prepareDives(const QString &tempfile, bool selected)
 			put_format(&mb, "</divelog>\n");
 		}
 		membuf = mb_cstring(&mb);
-		streamsize = strlen(membuf);
+		streamsize = mb.len - 1;
```

Comment:
This looks wrong to me.

mb_cstring() doesn't actually increase 'len' at all. It makes sure that there's enough _room_ for 'len', and adds a zero at the end, but the length remains the same. So the "mb.len - 1" is wrong. The length of the string is just "mb.len".

Put another way: mb_cstring() is very different from "put_char(0)". mb_cstring makes sure the membuffer contents can be used as a valid C string, *without modifying* the membuffer.

So "strlen(membuf)" should always just have been "mb.len". The two are the same (as long as the membuffer doesn't contain NUL characters anywhere else, of course).

Reply:
Oh, I see. It's the equivalent of `std::string::c_str()`. :-)

But looking closer - isn't the `mb_cstring` unnecessary?

The result is used in the line
```
               xmlDoc *doc = xmlReadMemory(membuf, streamsize, "divelog", NULL, 0);
```
which doesn't need NUL-termination, right?

Later the variable is overwritten with
```
                xmlDocDumpMemory(transformed, (xmlChar **)&membuf, &streamsize);
```

And it seems to me that `membuf` is not freed? Am I missing something?

You see, that is why *I* detest C-interfaces. I'm constantly struggling to understand whether the memory management is OK or not. In idiomatic C++ this just isn't an issue. I can't understand why someone would want to do this to themselves. :-P